Add busy indicator to mini reporter

jamestalmage opened the pull request:
Add a spinner to the mini reporter to indicate ongoing test progress.

Fixes #598.

jamestalmage commented:
![ava-spinner-demo](https://cloud.githubusercontent.com/assets/4082216/13512553/606304b0-e169-11e5-997f-ae61467af746.gif)

I ran all the visual tests and they all look good on iTerm2, OSX El Capitan.

sotojuan commented:
This is awesome, thanks!

sindresorhus commented on lib/reporters/mini.js:
```diff
@@ -18,22 +23,45 @@ function MiniReporter() {
+};
+
+MiniReporter.prototype.clearInterval = function () {
+	if (this.interval) {
```
Is there really any point in guarding this? `clearInterval` gracefully accepts anything.

sindresorhus commented on lib/reporters/mini.js:
```diff
@@ -18,22 +23,45 @@ function MiniReporter() {
+	var self = this;
+	this.interval = setInterval(function () {
+		self.spinnerFrame = (self.spinnerFrame + 1) % self.spinnerFrames.length;
+		self.write(self.spinnerChar() + self.currentTest);
```
Can you wrap the spinner char in `chalk.dim()`?

sindresorhus replied:
It's also a bit weird that `self.spinnerChar() + self.currentTest` is duplicated in `MiniReporter#test` too.

novemberborn replied:
Could abstract that into `writeStatusWithSpinner()` or something.

jamestalmage replied:
> Can you wrap the spinner char in chalk.dim()?

I just gave it a try, I can't tell the difference.

sindresorhus replied:
@jamestalmage It's not supported on all Terminals, but it dims the spinner to be less intruding. Could maybe also do `chalk.gray.dim()` to support all terminals.

sindresorhus commented:
Now the test title and the total count (71 passed) is vertically unaligned. They should be aligned.

sindresorhus commented on lib/reporters/mini.js:
```diff
@@ -18,22 +23,45 @@ function MiniReporter() {
 	this.currentStatus = '';
+	this.currentTest = '';
 	this.statusLineCount = 0;
+	this.spinnerFrame = 0;
```
`spinnerIndex` would be a better name. When it says `spinnerFrame` I thought it was the actual frame.

sindresorhus commented:
While you're at it, can you put a linebreak over the test title to give it some breathing-room?

novemberborn commented on lib/reporters/mini.js:
```diff
@@ -18,22 +23,45 @@ function MiniReporter() {
+	if (this.interval) {
+		clearInterval(this.interval);
+	}
+	this.interval = 0;
```
Reset to `null` since Node uses object refs?

novemberborn commented on lib/reporters/mini.js:
```diff
@@ -18,22 +23,45 @@ function MiniReporter() {
+
 MiniReporter.prototype.test = function (test) {
-	var status = '';
+	this.currentTest = this._test(test);
```
Should this be `currentStatus` instead?

jamestalmage replied:
`currentStatus` includes the spinner character. currentTest does not.

novemberborn replied:
Ah didn't realize there was already a `currentStatus` property.

novemberborn commented on lib/reporters/mini.js:
```diff
@@ -10,6 +11,10 @@ function MiniReporter() {
 
+	var spinnerDef = spinners.dots;
+	this.spinnerFrames = spinnerDef.frames;
+	this.spinnerInterval = spinnerDef.interval;
```
How about using `var dots = require('cli-spinners').dots` above and just referencing `dots.frames`/`dots.interval`?

jamestalmage replied:
@sindresorhus usually objects to that. I don't have a strong opinion either way.

jamestalmage replied:
Also, this way leaves us open to having a command line option to configure your spinner:

```
$ ava --spinner=balloon
```

then

```js
var spinnerDef = spinners[config.spinner];
```

jamestalmage commented:
Should pass / fail count be on the same line?

A: 

<img width="228" alt="screenshot 2016-03-05 21 40 58" src="https://cloud.githubusercontent.com/assets/4082216/13551977/fe8a169e-e31a-11e5-8a1f-d242831cf307.png">

vs.

B:

<img width="247" alt="screenshot 2016-03-05 21 40 00" src="https://cloud.githubusercontent.com/assets/4082216/13551980/04a29696-e31b-11e5-968c-6bd0f16670e9.png">

jamestalmage commented:
OK,
I incorporated most of the PR feedback.

Also, I switched it to option A as described above (counts on separate lines). That is the last commit, so it can be dropped if there is not agreement.

sindresorhus commented:
One more minor tweak. Move the spinner one char to the right without moving anything else. It's a bit too far from the test title and a bit too much crammed into the left edge.

`⠋  test title` => ` ⠋ test title`

sindresorhus commented:
I'm good with `A`, but will let @vdemedes chime in.

sindresorhus commented on lib/reporters/mini.js:
```diff
@@ -51,15 +85,18 @@ MiniReporter.prototype.test = function (test) {
-		status += '  ' + colors.error(this.failCount, 'failed');
+		if (this.passCount > 0) {
+			status += '\n';
+		}
```
Wouldn't you need to do this for `skipped` and `todo` tests results too?

jamestalmage replied:
Good call on `todo`. I implemented this before rebasing, so yeah - that almost certainly needs a fix.

Are skipped tests shown yet? I thought there was still a pending PR for that.

jamestalmage replied:
Oh - and I also missed everything in `final`

sindresorhus replied:
Yes, with master:

<img width="247" alt="screen shot 2016-03-06 at 14 49 01" src="https://cloud.githubusercontent.com/assets/170270/13552984/9cb8ac9c-e3aa-11e5-86fe-8198acafd3fb.png">

vdemedes commented:
I'm also ok with A (assuming all other counters will be updated too).

jamestalmage commented:
All changes made.

novemberborn commented on lib/reporters/mini.js:
```diff
@@ -18,10 +26,16 @@ function MiniReporter() {
+		self.spinnerIndex = (self.spinnerIndex + 1) % self.spinnerFrames.length;
+		self.write(self.prefix());
+	}, this.spinnerInterval);
+	return this.prefix('');
```
The `''` seems unnecessary here.

novemberborn commented:
Cool!

sindresorhus commented:
The default reporter looks so beautiful now 😍. Thanks James :)
